File: Excel/547 Challenge.py
import pandas as pd

path = "547 Sum and Diff both Triangular Numbers.xlsx"
test = pd.read_excel(path, usecols="A")

def generate_triangular_numbers(limit):
    return [n * (n + 1) // 2 for n in range(1, int((2 * limit) ** 0.5) + 1)]

def is_triangular_number(n):
    return ((8 * n + 1) ** 0.5).is_integer()

triangulars = generate_triangular_numbers(1000000)

triangulars_df = pd.DataFrame({'t1': triangulars})
result = triangulars_df.assign(key=0).merge(triangulars_df.assign(key=0), on='key').drop(columns='key')
result.columns = ['t1', 't2']

result = result[(result['t1'] < result['t2']) & 
                (result['t1'] + result['t2']).apply(is_triangular_number) & 
                (abs(result['t2'] - result['t1'])).apply(is_triangular_number) & 
                (result['t1'] != result['t2'])].reset_index(drop=True)

print(result)

File: Excel/500-599/547/547 Challenge.R
library(tidyverse)
library(readxl)

path = "Excel/547 Sum and Diff both Triangular Numbers.xlsx"
test  = read_excel(path, range = "A1:A21")

generate_triangular_numbers <- function(limit) {
  n <- 1
  triangulars <- c()
  
  while (TRUE) {
    t_n <- n * (n + 1) / 2
    if (t_n > limit) break
    triangulars <- c(triangulars, t_n)
    n <- n + 1
  }
  return(triangulars)
}

is_triangular_number <- function(n) {
  return((sqrt(8 * n + 1) - 1) %% 2 == 0)
}


tr = generate_triangular_numbers(1000000)

tr_pairs = expand.grid(tr, tr) %>%
  as_tibble() %>%
  filter(Var1 != Var2,
         Var1 < Var2,
         is_triangular_number(Var1 + Var2),
         is_triangular_number(abs(Var1 - Var2))) %>%
  arrange(Var1, Var2) %>%
  head(20) %>%
  unite("result", c(Var1, Var2), sep = ", ", remove = T)

print(tr_pairs)
